### PR TITLE
Added vlan_range support

### LINKED ISSUE #18: Add support to VLAN range
Original issue opened by @italovalcy at https://github.com/kytos/mef_eline/issues/241.

Starting on OpenFlow 1.2, many matching fields started to support the usage of field value + mask (sometimes called Bitwise match).  In other words, specific bits in the field must have specified values;  for example,  only source  IP addresses in a particular subnet (e.g., nw_src=10.1.0.0/255.255.0.0 means that the OF rule will match 10.1.0.1 and also 10.1.24.252, etc).

The VLAN_id field supports Mask and that can be used to define matches for "VLAN range". Accordingly, instead of having N OpenFlow rules for matching N vlans, range matches can be expressed as a collection of bitwise matches and end up with Z<=N flows (Z tends to be much lower than N, but the worst case they will be equal). For example, let's say that you want to create EVCs for VLANs from 1000 to 1023, the standard way of doing so is by creating 24 rules with matches including:

vlan_id=1000,
vlan_id=1001,
...
vlan_id=1023,

However, using bitwise matches, you can define this with only 2 flows:

vlan_id=0x03e8/0xfff8
vlan_id=0x03f0/0xfff0

### LINKED ISSUE #309: Update `_prepare_uni_flows` to deal with VLAN ranges
If the EVC contain a VLAN range, UNI flows must be split into several flows, one for each pair vlan/mask the range is divided.
The match of each flow should have vlan/mask.

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #18 
Closes #309 

### Summary

This PR needs [kytos](https://github.com/kytos-ng/kytos/pull/428) PR
Added support for `vlan_range` epic. Currently only when both UNIs have the same list of tags

### Local Tests

- Created, updated and deleted circuit.
- Restarted kytos to check consistency work. Partial updates for list of tags also tested. For example if both UNI have `[[20, 30]]` and they are being updated to `[[25, 30]]`.
- Added and updated tests.

### End-To-End Tests
```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.3.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 244 items

tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
tests/test_e2e_05_topology.py ..................                         [  8%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x................  [ 24%]
tests/test_e2e_11_mef_eline.py ......                                    [ 27%]
tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 30%]
tests/test_e2e_13_mef_eline.py ....Xs.s.....Xs.s.XXxX.xxxx..X........... [ 47%]
.                                                                        [ 47%]
tests/test_e2e_14_mef_eline.py x                                         [ 47%]
tests/test_e2e_15_mef_eline.py ....                                      [ 49%]
tests/test_e2e_20_flow_manager.py .....................                  [ 58%]
tests/test_e2e_21_flow_manager.py ...                                    [ 59%]
tests/test_e2e_22_flow_manager.py ...............                        [ 65%]
tests/test_e2e_23_flow_manager.py ..............                         [ 71%]
tests/test_e2e_30_of_lldp.py ....                                        [ 72%]
tests/test_e2e_31_of_lldp.py ...                                         [ 74%]
tests/test_e2e_32_of_lldp.py ...                                         [ 75%]
tests/test_e2e_40_sdntrace.py .............                              [ 80%]
tests/test_e2e_41_kytos_auth.py ........                                 [ 84%]
tests/test_e2e_42_sdntrace.py ..                                         [ 84%]
tests/test_e2e_50_maintenance.py ........................                [ 94%]
tests/test_e2e_60_of_multi_table.py .....                                [ 96%]
tests/test_e2e_70_kytos_stats.py ........                                [100%]
```

### CODE REVIEW

File: models/evc.py
Comment (Alopalao) [inline models/evc.py:1626]: This `return` is here to avoid early activations so consistency checks can be tested.

File: utils.py
Comment (Alopalao) [inline utils.py:100]: I cannot figure this case out, although I have not checked wireshark. For any flow with `vlan/mask` that has a 4095 mask, the string is replaced by an interger `vlan` when the flow is in the `switch`. For example `"101/4095"` is replaced by `101`.
The flow is sent correctly from `flow_manager`, [code](https://github.com/kytos-ng/flow_manager/blob/58f11e8680132dffc991e164de9894d663361423/main.py#L782), but received as string in `of_core`, [code](https://github.com/kytos-ng/of_core/blob/4503554180d51584e7adf5b0136e47a2fd8dddf1/main.py#L195) (check on reply.body).
For now, I am treating this case as a feature and not a bug since it does not interrupt any of `vlan_range` development. But probably not final.

**Review by viniarck — CHANGES_REQUESTED** [review]
@Alopalao, functionality-wise it's getting there. Excellent to this it working for the first time. I've hit an exception though, and also raised a few points. 

Regarding the all 1's mask I've also opened an issue to look into it, and then we can make a final decision about whether or not to send with all 1s mask.

**Comment by Alopalao** [comment]
Last commit is more stable. Tested with this [script](https://github.com/Alopalao/Alopalao_miscellaneous/blob/main/testing/create_evcs.py). This script work with the last update from [kytos](https://github.com/kytos-ng/kytos/pull/428), [topology](https://github.com/kytos-ng/topology/pull/175) and [of_lldp](https://github.com/kytos-ng/of_lldp/pull/99) PRs.
It runs as `python3 evcs.py 5`. It set tag ranges to `"01:1"` and `"02:1"` interfaces and creates a set number of circuits.
The result should an empty `available_tags["vlan"]` for `"01:1"` and `"02:1"` interfaces.

File: models/evc.py
Comment (Alopalao) [reply models/evc.py:1379]: Fixed this, I made a mistake with the traces index.

File: models/evc.py
Comment (Alopalao) [reply models/evc.py:303]: Simplified.

File: main.py
Comment (Alopalao) [reply main.py]: Omitted them all.

File: main.py
Comment (Alopalao) [reply main.py]: Deleted.

File: main.py
Comment (Alopalao) [reply main.py:707]: I totally forgot about `"untagged"` and `"any"`. Probably going to add an variable for each interface `special_vlan = {"vlan": {"untagged", "any"}}`. But not addressed yet.

File: main.py
Comment (Alopalao) [reply main.py]: Deleted logs

File: main.py
Comment (Alopalao) [reply main.py:373]: Forgot about this exception,  now it is handled correctly.

File: main.py
Comment (Alopalao) [reply main.py:287]: Fixed.

File: db/models.py
Comment (Alopalao) [reply db/models.py]: Added.

File: requirements/dev.in
Comment (Alopalao) [inline requirements/dev.in:8]: Reverting this when PR lands

File: main.py
Comment (Alopalao) [reply main.py]: I reverted this changed. Adding `special_vlan = {"vlan": {"untagged", "any"}}` could be done in another PR.

File: db/models.py
Comment (viniarck) [reply db/models.py]: Looks good.

File: main.py
Comment (viniarck) [reply main.py]: Right. I'm glad you've also caught this case. It's fine to OK to keep `_is_duplicated_evc(evc)` in the meantime, and then if you have another better suggestion to do in another issue/PR. 

Regarding `_is_duplicated_evc(evc)` looks like it's not being called during an update if you could also explore to see if this case is also being validated during an update, that'd be great.

File: models/evc.py
Comment (viniarck) [reply models/evc.py:1616]: @Alopalao as we spoke in the meeting, if you could `git rebase` with this branch `fix/handle_switch_changes` of https://github.com/kytos-ng/mef_eline/pull/383, that'd be great. That way the linter error would also be gone. It was great how you managed to progress in the meantime.

**Review by viniarck — COMMENTED** [review]
@Alopalao, great PR. 

I've just done another pass closing the threads had I had opened. Fantastic to also see how you explored locally, and the new e2e that you're shipping. I'll do a final review once you also rebase with PR #383, let me know when you need a final review.

**Comment by viniarck** [comment]
Changelog also hasn't been updated

File: models/evc.py
Comment (Alopalao) [reply models/evc.py]: Rebased it and tested it, both are working well.

File: main.py
Comment (Alopalao) [reply main.py]: Added `_is_duplicated_evc(evc)` to `update()`. Also I modified it so it works with updates.
Commit [1086884a3906b023f81aa932ec2cde66f5652920](https://github.com/kytos-ng/mef_eline/pull/396/commits/1086884a3906b023f81aa932ec2cde66f5652920)

**Comment by Alopalao** [comment]
Bypassing checking of `tags` for `use_tags()` and `make_tags_available()` since these tags are not managed by the user.
Commit [a29489619e26a6ebea63ad69ce6c386992644dfc](https://github.com/kytos-ng/mef_eline/pull/396/commits/a29489619e26a6ebea63ad69ce6c386992644dfc)

File: models/evc.py
Comment (viniarck) [reply models/evc.py]: Great. Certain `master` originated commits here that haven't been merged into #383 is making the review slightly more difficult, if you could talk to @Ktmi to merge `master` into his branch, that'd be great.

File: utils.py
Comment (viniarck) [inline utils.py:128]: `mask_list` currently can be initialized as `None` so this iteration here or in other places, in some cases it might crash. So, on Kytos core we might want to initialize as an empty list if the initial `mask_list` was `None`.

File: models/path.py
Comment (viniarck) [inline models/path.py:48]: In places when you need to pass a boolean in the args if you could pass `check_order=False` explicitly that'd improve readability, without having to check the implementation source code to try to remember

File: models/evc.py
Comment (viniarck) [inline models/evc.py:457]: Can you double this `False` arg here, I believe `use_lock=True` should still be in place here, I'm not sure if you also meant to use `check_order`

Comment (Alopalao) [reply models/evc.py]: Yes, you are correct, my mistake. Fixed.

File: models/path.py
Comment (Alopalao) [reply models/path.py]: Added it

File: utils.py
Comment (Alopalao) [reply utils.py]: `mask_list` initialize value change in last kytos [commit](https://github.com/kytos-ng/kytos/pull/428/commits/f80d202828284e2a229937ad1eba7356edf4a049#).

File: models/evc.py
Comment (viniarck) [reply models/evc.py]: @Alopalao, I was checking David's PR commits history he's already got the correct commits there:

```
commit c9d91ed51799ce56e76e937d92bcb9cae6e1a9fb
Author: Vinicius Arcanjo <vindasil@fiu.edu>
Date:   Tue Oct 31 17:50:48 2023 -0300

    feat: add set_vlan only if in_vlan_a != in_vlan_z when pushing UNI flow
    related to issue 389


```

and on `master` same sha1:

```
commit c9d91ed51799ce56e76e937d92bcb9cae6e1a9fb
Author: Vinicius Arcanjo <vindasil@fiu.edu>
Date:   Tue Oct 31 17:50:48 2023 -0300

    feat: add set_vlan only if in_vlan_a != in_vlan_z when pushing UNI flow
    related to issue 389
```

So you'll need to sort out why these commits got rewritten, if you need help with this let me know.

File: main.py
Comment (viniarck) [inline main.py:927]: Trivial detail but you can group exception types since you're using the same code in the except claus, you can turn the prior except as `except (ValueError, KytosTagError) as exception:`, that way eliminating the need to repeat the `log.error` and return None statement

File: models/evc.py
Comment (viniarck) [inline models/evc.py:1664]: Looks like the `# TODO: Remove when addressing issue #369` got accidentally removed if you could add it again

File: CHANGELOG.rst
Comment (viniarck) [inline CHANGELOG.rst:14]: If you could also mention in the changelog that the openapi has been updated accordingly and for more information users should check it out, that'd be great.

File: models/evc.py
Comment (viniarck) [inline models/evc.py:1458]: Up to this point don't we have validation guarantees during EVC creation and updates that all EVCs will have symmetric mask_list? If this requirement ever change I can see this becoming difficult to remember and identify, even though in some cases it'll result in index error. Let me know if you're seeing any other case or had any other reason to compute the length with this `or`. If you want extra guarantees here, maybe if the lengths are diff then you raise an exception just so it's immediately clear that's not supported?

Comment (Alopalao) [reply models/evc.py:1458]: Added this because I was creating TAGs in kytos console with empty `mask_list`s to test exceptions. Currently without any modification to an EVC, both `value` and `mask_list` should be the same so this `len() or len()` is not should not be needed. Going to delete it.

Comment (viniarck) [reply models/evc.py:1458]: Got it. OK.

Comment (Alopalao) [reply models/evc.py:1458]: Updated, commit [918e21324fdd2dc3d6615cf967613ae75802fb31](https://github.com/kytos-ng/mef_eline/pull/407/commits/918e21324fdd2dc3d6615cf967613ae75802fb31)

File: CHANGELOG.rst
Comment (Alopalao) [reply CHANGELOG.rst:14]: Updated changelog, commit [918e21324fdd2dc3d6615cf967613ae75802fb31](https://github.com/kytos-ng/mef_eline/pull/407/commits/918e21324fdd2dc3d6615cf967613ae75802fb31)

File: main.py
Comment (Alopalao) [reply main.py:927]: Got it, grouped similar cases. Commit [918e21324fdd2dc3d6615cf967613ae75802fb31](https://github.com/kytos-ng/mef_eline/pull/407/commits/918e21324fdd2dc3d6615cf967613ae75802fb31)

File: models/evc.py
Comment (Alopalao) [reply models/evc.py:1458]: Made a mistake with length, fixed commit [1613b6f110a123dca424d374c8d7ea735be13bb3](https://github.com/kytos-ng/mef_eline/pull/407/commits/1613b6f110a123dca424d374c8d7ea735be13bb3)

**Comment by viniarck** [comment]
Closing this since Aldo's PR #407 has landed. Nicely done, Aldo.

**Pull request closed without merging**
